cli/src/cmd/forge/script/broadcast.rs

```diff
@@ -279,43 +279,105 @@ impl ScriptArgs {
+            // before broadcasting.
+            let per_gas = if let Some(gas_price) = self.with_gas_price {
+                gas_price
+            } else {
+                match new_txes.front().unwrap().typed_tx() {
+                    TypedTransaction::Legacy(_) | TypedTransaction::Eip2930(_) => {
+                        provider.get_gas_price().await?
+                    }
+                    TypedTransaction::Eip1559(_) => provider.estimate_eip1559_fees(None).await?.0,
                 }
-                TypedTransaction::Eip1559(_) => provider.estimate_eip1559_fees(None).await?.0,
-            }
-        };
+            };
 
-        println!("\n==========================");
-        println!("\nEstimated total gas used for script: {}", total_gas);
-        println!(
-            "\nEstimated amount required: {} ETH",
-            format_units(total_gas.saturating_mul(per_gas), 18)
-                .unwrap_or_else(|_| "[Could not calculate]".to_string())
-                .trim_end_matches('0')
-        );
-        println!("\n==========================");
+            println!("\n==========================");
+            println!("\nEstimated total gas used for script: {}", total_gas);
+            println!(
+                "\nEstimated amount required: {} ETH",
+                format_units(total_gas.saturating_mul(per_gas), 18)
+                    .unwrap_or_else(|_| "[Could not calculate]".to_string())
+                    .trim_end_matches('0')
+            );
+            println!("\n==========================");
+        }
         Ok(new_txes)
     }
+    /// Uses the signer to submit a transaction to the network. If it fails, it tries to retrieve
+    /// the transaction hash that can be used on a later run with `--resume`.
+    async fn broadcast<T, U>(
+        &self,
+        signer: &SignerMiddleware<T, U>,
+        mut legacy_or_1559: TypedTransaction,
+    ) -> Result<TxHash, BroadcastError>
+    where
+        T: Middleware,
+        U: Signer,
+    {
+        tracing::debug!("sending transaction: {:?}", legacy_or_1559);
+
+        // Chains which use `eth_estimateGas` are being sent sequentially and require their gas to
+        // be re-estimated right before broadcasting.
+        if has_different_gas_calc(signer.signer().chain_id()) || self.skip_simulation {
+            // if already set, some RPC endpoints might simply return the gas value that is already
+            // set in the request and omit the estimate altogether, so we remove it here
+            let _ = legacy_or_1559.gas_mut().take();
+
+            self.estimate_gas(&mut legacy_or_1559, signer.provider()).await?;
+        }
+
+        // Signing manually so we skip `fill_transaction` and its `eth_createAccessList` request.
+        let signature = signer
+            .sign_transaction(
+                &legacy_or_1559,
+                *legacy_or_1559.from().expect("Tx should have a `from`."),
```

Comment:
is this caused by user error or by us? if by user error would prefer we return an `Error` instead of panic

Reply:
I just moved this part of the code around, so I'm not exactly sure in what scenario this panic would be thrown. However since this is very late in the "script lifecycle" (after off-chain / on-chain simulations and just before broadcasting the tx), I'd assume that a user error would have been caught earlier.